source/DamageProfile.cpp

```diff
@@ -0,0 +1,155 @@
+	}
+
+	// Instantaneous damage types.
+	// Energy, heat, and fuel damage are blocked 50% by shields.
+	// Hull damage is blocked 100%.
+	// Shield damage is blocked 0%.
+	damage.shieldDamage *= shieldFraction;
+	damage.hullDamage = (weapon.HullDamage()
+		+ weapon.RelativeHullDamage() * attributes.Get("hull"))
+		* ScaleType(1., attributes.Get("hull protection"));
+	damage.energyDamage = (weapon.EnergyDamage()
+		+ weapon.RelativeEnergyDamage() * attributes.Get("energy capacity"))
+		* ScaleType(.5, attributes.Get("energy protection"));
+	damage.heatDamage = (weapon.HeatDamage()
+		+ weapon.RelativeHeatDamage() * ship.MaximumHeat())
+		* ScaleType(.5, attributes.Get("heat protection"));
+	damage.fuelDamage = (weapon.FuelDamage()
+		+ weapon.RelativeFuelDamage() * attributes.Get("fuel capacity"))
+		* ScaleType(.5, attributes.Get("fuel protection"));
+
+	// DoT damage types with an instantaneous analog.
+	// Ion and burn damage are blocked 50% by shields.
+	// Corrosion and leak damage are blocked 100%.
+	// Discharge damage is blocked 0%.
+	damage.dischargeDamage = weapon.DischargeDamage() * ScaleType(0., attributes.Get("discharge protection"));
+	damage.corrosionDamage = weapon.CorrosionDamage() * ScaleType(1., attributes.Get("corrosion protection"));
+	damage.ionDamage = weapon.IonDamage() * ScaleType(.5, attributes.Get("ion protection"));
+	damage.burnDamage = weapon.BurnDamage() * ScaleType(.5, attributes.Get("burn protection"));
+	damage.leakDamage = weapon.LeakDamage() * ScaleType(1., attributes.Get("leak protection"));
+
+	// Unique special damage types.
+	// Disruption and slowing are blocked 50% by shields.
+	damage.disruptionDamage = weapon.DisruptionDamage() * ScaleType(.5, attributes.Get("disruption protection"));
+	damage.slowingDamage = weapon.SlowingDamage() * ScaleType(.5, attributes.Get("slowing protection"));
+
+	// Hit force is blocked 0% by shields.
+	double hitForce = weapon.HitForce() * ScaleType(0., attributes.Get("force protection"));
+	if(hitForce)
+	{
+		Point d = ship.Position() - position;
+		double distance = d.Length();
+		if(distance)
+			damage.forcePoint = (hitForce / distance) * d;
+	}
+
+	return damage;
+}
+
+
+
+// Finish any calculations that were started in the constructor.
+void DamageProfile::FinishPrecalculations(const Ship &ship, DamageProfile::DamageDealt &damage) const
+{
+	// Finish the blast radius calculations.
+	if(isBlast && weapon.IsDamageScaled())
+	{
+		// Rather than exactly compute the distance between the explosion and
+		// the closest point on the ship, estimate it using the mask's Radius.
+		double d = max(0., (position - ship.Position()).Length() - ship.GetMask().Radius());
+		double finalR = d * d * rSquared;
+		damage.scaling *= k / ((1. + finalR * finalR) * (1. + finalR * finalR));
+	}
+	// If damage falloff scaling was skipped before, compute it now.
+	if(skipFalloff && weapon.HasDamageDropoff())
+		damage.scaling *= weapon.DamageDropoff(distanceTraveled);
```

Comment:
Why not move this calculation to the SetDistance function? (That seems to be a very natural place to perform this type of calculation.)

Reply:
Because the damage object isn't created yet in SetDistance.

Reply:
Still, the function name (`FinishPrecalculations`) and the comment above the function make this whole function feel a bit unnatural to me. It suggests that there were some calculations for which there was no proper place to perform them, so they were just put in this function.
(Maybe it helps to make the function name more descriptive of the calculations that are done in the function.)

Reply:
That's admittedly a placeholder name.